flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java

```diff
@@ -70,28 +104,68 @@ public ActorGateway getArchiveGateway() {
 		return archiveGateway;
 	}
 
+	/**
+	 * Returns the current redirect address or <code>null</code> if the job manager associated with
+	 * this web monitor is leading. In that case, work with the gateway directly.
+	 */
+	public String getRedirectAddress() {
+		return redirectWebMonitorAddress;
+	}
 
 	@Override
 	public void notifyLeaderAddress(String leaderAddress, UUID leaderSessionID) {
 		if (leaderAddress != null && !leaderAddress.equals("")) {
 			try {
-				ActorRef jobManager = AkkaUtils.getActorRef(
-						leaderAddress,
-						actorSystem,
+				ActorRef jobManager = AkkaUtils.getActorRef(leaderAddress, actorSystem,
```

Comment:
This is a blocking call which will block the `NodeCache's` thread which calls the `NodeCacheListener`. This should be avoided by using `AkkaUtils.getActorRefFuture` and a `Promise`.